jerry-core/ecma/operations/ecma-container-object.c

```diff
@@ -614,12 +620,203 @@ ecma_op_container_delete (ecma_value_t this_arg, /**< this argument */
+    }
+
+    ecma_value_t put_result = ecma_op_object_put (obj_p,
+                                                  prop_name_p,
+                                                  ecma_make_uint32_value (index),
+                                                  true);
+
+    JERRY_ASSERT (ecma_is_value_true (put_result));
+
+    ecma_free_value (index_value);
+  }
+  else
+  {
+    ext_obj_p->u.pseudo_array.u1.iterator_index++;
+  }
+
+  if (index >= length)
+  {
+    ECMA_SET_POINTER (ext_obj_p->u.pseudo_array.u2.iterated_value_cp, NULL);
+    ecma_free_values_collection (props_p, 0);
+    return ecma_create_iter_result_object (ECMA_VALUE_UNDEFINED, ECMA_VALUE_TRUE);
+  }
+
+  uint8_t iterator_type = ext_obj_p->u.pseudo_array.extra_info;
+
+  ecma_value_t *ecma_value_p = ecma_collection_iterator_init (props_p);
+
+  ecma_value_t ret_value = ECMA_VALUE_UNDEFINED;
+
+  while (ecma_value_p != NULL)
+  {
+    if (index > 0)
+    {
+      index--;
+      ecma_value_p = ecma_collection_iterator_next (ecma_value_p);
+      continue;
+    }
+
+    ecma_string_t *prop_name_p = ecma_get_prop_name_from_value (*ecma_value_p);
+    ecma_property_t *property_p = ecma_find_named_property (internal_obj_p, prop_name_p);
+    JERRY_ASSERT (property_p != NULL);
+
+    if (ecma_is_value_empty (ECMA_PROPERTY_VALUE_PTR (property_p)->value))
+    {
+      ecma_value_p = ecma_collection_iterator_next (ecma_value_p);
+
+      if (ecma_value_p == NULL)
+      {
+        ret_value = ecma_create_iter_result_object (ECMA_VALUE_UNDEFINED, ECMA_VALUE_TRUE);
+      }
+      continue;
+    }
+
+    ecma_value_t value = ecma_copy_value (ECMA_PROPERTY_VALUE_PTR (property_p)->value);
+    ecma_value_t key_arg;
+
+    if (ecma_prop_name_is_map_key (prop_name_p))
+    {
+      key_arg = prop_name_p->u.value;
+    }
+    else
+    {
+      if (ECMA_IS_DIRECT_STRING (prop_name_p)
+          && ECMA_GET_DIRECT_STRING_TYPE (prop_name_p) == ECMA_DIRECT_STRING_ECMA_INTEGER)
+      {
+        key_arg = ecma_make_uint32_value ((uint32_t) ECMA_GET_DIRECT_STRING_VALUE (prop_name_p));
+      }
+      else
+      {
+        key_arg = *ecma_value_p;
+      }
+    }
+
+    if (iterator_type == ECMA_ITERATOR_KEYS)
+    {
+      ret_value = ecma_create_iter_result_object (is_set ? value : key_arg, ECMA_VALUE_FALSE);
+    }
+    else if (iterator_type == ECMA_ITERATOR_VALUES)
+    {
+      ret_value = ecma_create_iter_result_object (value, ECMA_VALUE_FALSE);
+    }
+    else
+    {
+      JERRY_ASSERT (iterator_type == ECMA_ITERATOR_KEYS_VALUES);
+
+      ecma_value_t entry_array_value;
+      entry_array_value = ecma_create_array_from_iter_element (value, is_set ? value : key_arg);
+
+      ret_value = ecma_create_iter_result_object (entry_array_value, ECMA_VALUE_FALSE);
+      ecma_free_value (entry_array_value);
+    }
+
+    ecma_free_value (value);
+    break;
+  }
+
+  ecma_free_values_collection (props_p, 0);
+
+  return ret_value;
+} /* ecma_op_container_iterator_next */
```

Comment:
I think the guards of this file are not correct. What happens if only `JERRY_ES2015_BUILTIN_SET` is enabled? What happens if `JERRY_ES2015_BUILTIN_MAP` and `JERRY_ES2015_BUILTIN_SET` are enabled, but `JERRY_ES2015_BUILTIN_ITERATOR` is disabled?

Reply:
`JERRY_ES2015_BUILTIN_SET` requires the `JERRY_ES2015_BUILTIN_MAP` so it cannot cause any problem. However, your second thought is currently a valid problem I'll check it.

Reply:
Fixed. Now the set is independent from the map builtin, also the corresponding iterator related routines are guarded correctly.